src/main/java/com/zendesk/maxwell/row/RowIdentity.java

```diff
@@ -0,0 +1,107 @@
+package com.zendesk.maxwell.row;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.*;
+
+public class RowIdentity implements Serializable {
+	private String database;
+	private String table;
+	private final List<Map.Entry<String, Object>> primaryKeyColumns;
+
+	// slightly less verbose pair construction
+	public static Map.Entry<String, Object> pair(String key, Object value) {
+		return new AbstractMap.SimpleImmutableEntry<>(key, value);
+	}
+
+	public RowIdentity(String database, String table, List<Map.Entry<String, Object>> primaryKeyColumns) {
+		this.database = database;
+		this.table = table;
+		this.primaryKeyColumns = primaryKeyColumns;
+	}
+
+	public String getDatabase() {
+		return database;
+	}
+
+	public String getTable() {
+		return table;
+	}
+
+	public String toKeyJson(RowMap.KeyFormat keyFormat) throws IOException {
+		MaxwellJson json = MaxwellJson.getInstance();
+		JsonGenerator g = json.reset();
+		if ( keyFormat == RowMap.KeyFormat.HASH )
+			writeKeyJsonHash(g);
+		else
+			writeKeyJsonArray(g);
+		return json.consume();
+	}
+
+	private void writeKeyJsonHash(JsonGenerator g) throws IOException {
+		writeStartCommon(g);
+		if (primaryKeyColumns.isEmpty()) {
+			g.writeStringField(FieldNames.UUID, UUID.randomUUID().toString());
+		} else {
+			for (Map.Entry<String,Object> pk : primaryKeyColumns) {
+				writePrimaryKey(g, "pk." + pk.getKey().toLowerCase(), pk.getValue());
```

Comment:
same note here, if you've got a one-line function that's called exactly once you're doing something off...

Reply:
This was because I pulled it out of #1176, so it will eventually be used multiple times. But I've inlined it for now.